Chapter3/homework/5

littleNoodle opened the pull request.

littleNoodle committed:
update MostCommonLetterFinder.cs

littleNoodle committed:
commit

littleNoodle committed:
add Clean class

littleNoodle committed:
commit

littleNoodle committed:
fix bug

littleNoodle committed:
fix bug twice

Almantask commented on Src/BootCamp.Chapter/ToggleableGridJagged.cs:
```diff
@@ -1,18 +1,52 @@
+            _toggle[x][y] = !_toggle[x][y];
+            StringBuilder sb = new StringBuilder();            
+            CheckInput(x, y);
+            for(int i = 0; i < _toggle.GetLength(0); i++)
+            {
+                for (int j = 0; j < _toggle[i].Length; j++)
+                {
+                    if (_toggle[i][j])
+                    {
+                        sb.Append("■");
+                    }
+                    else
+                    {
+                        sb.Append(" ");
+                    }
+                }
+                sb.Append(Environment.NewLine);
+            }           
+            var str = sb.ToString().TrimEnd("\r\n".ToCharArray());
```
Split to a function of its own

Almantask replied:
So the function will encapsulate sb- it's a low level detail

Almantask replied:
Avoid mixing high level functions with implementation of other functions.

littleNoodle replied:
Sorry for late response.
For this issue, I regard 2D array as special jagged array and define DisplayGrid class to reuse.
Also spend some time to solve the null reference exception about how to make 2D array works.
Through this I could handle 2D array initialization

Almantask commented on Src/BootCamp.Chapter/ToggleableGridJagged.cs:
```diff
@@ -1,18 +1,52 @@
         public void Toggle(int x, int y)
         {
+            _toggle[x][y] = !_toggle[x][y];
+            StringBuilder sb = new StringBuilder();            
```
Move this close to where it's used- next to loop

littleNoodle replied:
I leave the line 19 at Toggle function and make Display function only judge element value( true/false) of array.
High/Low level function is a great strategy, I should focus on code reuse. 
Have a good day~

Almantask reviewed (changes requested):
Try splitting the rendering of a grid from a logical representation and triggering toggles. When you couple both, you have quite a lot of duplication

littleNoodle committed:
modify-1

littleNoodle committed:
commit

#1 regard 2D array as special jagged array to reuse code
#2 add DisplayGrid class
#3 sovle Null Reference exception

Almantask commented:
Please reply to remarks by telling how you resolved them.

littleNoodle commented:
> Please reply to remarks by telling how you resolved them.

littleNoodle commented:
> Please reply to remarks by telling how you resolved them.
I add the reponse above.
Happy new year !
